### PR TITLE
chore: integrate msw to browser to mock api requests

### LINKED ISSUE #14395: [Task]: mock API calls/response for RBAC
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

> As of now we are just consuming the mock data directly by keeping it in a variable and just setting it to the state on load, we are not getting it async in an effect. So if we could mock the process using a mock response data in a json file, once the API is done we can just replace the path to json file with path to url endpoint.

- mock API calls
- mock response data for RBAC 
- create sagas and actions

### PR DESCRIPTION
**Pull request by berzerkeer**
## Description

As of now, when the backend api's are not ready we mock api data by returning a JSON in promise or just directly using it. There is no way to  mock the `serve` part unless you put the json files in a public folder and fetch using axios or fetch. So `msw` is made to solve that, we are already using it for our test env and now I am adding it to browser env.  We only use it in `development` and the `serviceworker` also only gets registered in the dev environment. 

Fixes #14395

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Manual

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/public/mockServiceWorker.js [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    }
+  }
+})
+
+// Resolve the "master" client for the given event.
+// Client that issues a request doesn't necessarily equal the client
+// that registered the worker. It's with the latter the worker should
+// communicate with during the response resolving phase.
+async function resolveMasterClient(event) {
+  const client = await self.clients.get(event.clientId)
```
Comment (IAmAnubhavSaini) [inline app/client/public/mockServiceWorker.js:90]: If you are awaiting, then you need to handle for error via try/catch.

File: app/client/public/mockServiceWorker.js [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+  const client = await resolveMasterClient(event)
+  const response = await getResponse(event, client, requestId)
+
+  // Send back the response clone for the "response:*" life-cycle events.
+  // Ensure MSW is active and ready to handle the message, otherwise
+  // this message will pend indefinitely.
+  if (activeClientIds.has(client.id)) {
+    ;(async function () {
+      const clonedResponse = response.clone()
+      sendToClient(client, {
```
Comment (IAmAnubhavSaini) [inline app/client/public/mockServiceWorker.js:120]: 1. You are returning `Promise` from that function. Need `await` here.

File: app/client/public/mockServiceWorker.js [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+async function handleRequest(event, requestId) {
+  const client = await resolveMasterClient(event)
+  const response = await getResponse(event, client, requestId)
+
+  // Send back the response clone for the "response:*" life-cycle events.
+  // Ensure MSW is active and ready to handle the message, otherwise
+  // this message will pend indefinitely.
+  if (activeClientIds.has(client.id)) {
+    ;(async function () {
```
Comment (IAmAnubhavSaini) [inline app/client/public/mockServiceWorker.js:118]: 2. This doesn't need `;`.
3. Need await here. An async function, even IIAFE, should be awaited.

File: app/client/public/mockServiceWorker.js [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+function respondWithMock(clientMessage) {
+  return new Response(clientMessage.payload.body, {
+    ...clientMessage.payload,
+    headers: clientMessage.payload.headers,
+  })
+}
+
+function uuidv4() {
+  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
+    const r = (Math.random() * 16) | 0
```
Comment (IAmAnubhavSaini) [inline app/client/public/mockServiceWorker.js:318]: Use crypto API.

https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API

Comment (IAmAnubhavSaini) [reply app/client/public/mockServiceWorker.js:318]: https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID

**Review by IAmAnubhavSaini — CHANGES_REQUESTED** [review]

**Comment by berzerkeer** [comment]
@IAmAnubhavSaini the file `mockServiceWorker.js` is auto generated by msw. I didn't create it.

**Review by IAmAnubhavSaini — DISMISSED** [review]

**Review by ankitakinger — APPROVED** [review]

**Pull request merged**